### ISSUE TITLE
Iteration stopped

### ISSUE THREAD

**Issue by lehahoang** [Hananel-Hazan/bindsnet#188]
Hi,
I have encountered an issue of running the file "eth_mnist.py". 
I used a loop with three iterations to train the network three times. It worked with the small number of training set. However, the training process stopped right after the first iteration when 60000 training set was selected. The attached figure depicts the message from the shell.
![screenshot at 2019-02-12 10-10-32](https://user-images.githubusercontent.com/37591808/52624608-68910600-2eaf-11e9-9873-6ff958de96ed.png)
The issue might come from this syntax below.
sample = next(data_loader)
Regarding to the setup, I changed the number of excitatory neurons up to 400.
Is there anyone who has the same issue like me?
Thank you.
BR

**Comment by lehahoang** [comment]
Hi,
I have looked into my code. I guess that after the first iteration of the outer loop, the pointer of the "next(data_loader)" pointed to the last dataset (60000th). Hence, there was no dataset to be loaded for the second training process. I assume that we have to call "data_loader = poisson_loader(data=images, time=time, dt=dt)" syntax again to get the "new" training set.
That's why now my code look absolutely crazy:
##........
if not train:
    update_interval = n_test
for j in range(epoch):
    n_sqrt = int(np.ceil(np.sqrt(n_neurons)))
    start_intensity = intensity

    # Build network.
    network = DiehlAndCook2015(n_inpt=784, n_neurons=n_neurons, exc=exc, inh=inh, dt=dt, norm=78.4, theta_plus=1)

    # Voltage recording for excitatory and inhibitory layers.
    exc_voltage_monitor = Monitor(network.layers['Ae'], ['v'], time=time)
    inh_voltage_monitor = Monitor(network.layers['Ai'], ['v'], time=time)
    network.add_monitor(exc_voltage_monitor, name='exc_voltage')
    network.add_monitor(inh_voltage_monitor, name='inh_voltage')

    # Load MNIST data.
    images, labels = MNIST(path=os.path.join('..', '..', 'data', 'MNIST'), download=True).get_train()
    images = images.view(-1, 784)
    images *= intensity

    # Lazily encode data as Poisson spike trains.
    data_loader = poisson_loader(data=images, time=time, dt=dt)

    # Record spikes during the simulation.
    spike_record = torch.zeros(update_interval, time, n_neurons)

    # Neuron assignments and spike proportions.
    assignments = -torch.ones_like(torch.Tensor(n_neurons))
    proportions = torch.zeros_like(torch.Tensor(n_neurons, 10))
    rates = torch.zeros_like(torch.Tensor(n_neurons, 10))

    # Sequence of accuracy estimates.
    accuracy = {'all': [], 'proportion': []}

    spikes = {}
    for layer in set(network.layers) - {'X'}:
        spikes[layer] = Monitor(network.layers[layer], state_vars=['s'], time=time)
        network.add_monitor(spikes[layer], name='%s_spikes' % layer)

    # Train the network.
    print('\nBegin training.\n')
    start_time = t()
    start = t()
    print('This is the ', j+1, 'th epoch')
    for i in range(n_train):
        if i % progress_interval == 0:
            print('Progress: %d / %d (%.4f seconds)' % (i, n_train, t() - start))
            start = t()

        if i % update_interval == 0 and i > 0:
            # Get network predictions.
            all_activity_pred = all_activity(spike_record, assignments, 10)
            proportion_pred = proportion_weighting(spike_record, assignments, proportions, 10)

            # Compute network accuracy according to available classification strategies.
            accuracy['all'].append(100 * torch.sum(labels[i - update_interval:i].long() \
                                                   == all_activity_pred) / update_interval)
            accuracy['proportion'].append(100 * torch.sum(labels[i - update_interval:i].long() \
                                                          == proportion_pred) / update_interval)

            print('\nAll activity accuracy: %.2f (last), %.2f (average), %.2f (best)' \
                  % (accuracy['all'][-1], np.mean(accuracy['all']), np.max(accuracy['all'])))
            print('Proportion weighting accuracy: %.2f (last), %.2f (average), %.2f (best)\n' \
                  % (accuracy['proportion'][-1], np.mean(accuracy['proportion']),
                     np.max(accuracy['proportion'])))

            # Assign labels to excitatory layer neurons.
            assignments, proportions, rates = assign_labels(spike_record, labels[i - update_interval:i], 10, rates)

        # Get next input sample.
        sample = next(data_loader)
        inpts = {'X': sample}

        # Run the network on the input.
        network.run(inpts=inpts, time=time)

        # Get voltage recording.
        exc_voltages = exc_voltage_monitor.get('v')
        inh_voltages = inh_voltage_monitor.get('v')

        # Add to spikes recording.
        spike_record[i % update_interval] = spikes['Ae'].get('s').t()

        # Optionally plot various simulation information.
        if plot:
            inpt = inpts['X'].view(time, 784).sum(0).view(28, 28)
            input_exc_weights = network.connections[('X', 'Ae')].w
            square_weights = get_square_weights(input_exc_weights.view(784, n_neurons), n_sqrt, 28)
            square_assignments = get_square_assignments(assignments, n_sqrt)
            voltages = {'Ae': exc_voltages, 'Ai': inh_voltages}

            if i == 0:
                #inpt_axes, inpt_ims = plot_input(images[i].view(28, 28), inpt, label=labels[i])
                #spike_ims, spike_axes = plot_spikes({layer: spikes[layer].get('s') for layer in spikes})
                weights_im = plot_weights(square_weights)
                assigns_im = plot_assignments(square_assignments)
                perf_ax = plot_performance(accuracy)
                #voltage_ims, voltage_axes = plot_voltages(voltages)

            else:
                #inpt_axes, inpt_ims = plot_input(images[i].view(28, 28), inpt, label=labels[i], axes=inpt_axes,
                #                                 ims=inpt_ims)
                #spike_ims, spike_axes = plot_spikes({layer: spikes[layer].get('s') for layer in spikes},
                #                                    ims=spike_ims, axes=spike_axes)
                weights_im = plot_weights(square_weights, im=weights_im)
                assigns_im = plot_assignments(square_assignments, im=assigns_im)
                perf_ax = plot_performance(accuracy, ax=perf_ax)
                #voltage_ims, voltage_axes = plot_voltages(voltages, ims=voltage_ims, axes=voltage_axes)
            plt.pause(1e-8)
        if i==n_train-1:
            network.save('./trained_eth_network_2.pt')
            print('network saved')
            network.reset_()  # Reset state variables.
        else:
            network.reset_()  # Reset state variables.
    print('Finish the ', j+1, 'th epoch')

print('Progress: %d / %d (%.4f seconds)\n' % (n_train, n_train, t() - start))
print('Training complete.\n')
time_execution = (t() - start_time)/3600
print('Total time for training is {:f} hours\n'.format(time_execution))

**Comment by dee0512** [comment]
Hi @lehahoang,

This example was meant to run only one epoch. However, it should be easy to modify the code for more than one epochs as follows:
```
for j in range(epochs):
     for datum in data_loader:
```

**Comment by lehahoang** [comment]
Hi @dee0512 ,
Thank you for your suggestion. I will try it.
